sklearn/base.py

```diff
@@ -597,6 +605,134 @@ def _validate_data(
 
         return out
 
+    def _set_callbacks(self, callbacks):
+        """Set callbacks for the estimator.
+
+        Parameters
+        ----------
+        callbacks : callback or list of callbacks
+            the callbacks to set.
+        """
+        if not isinstance(callbacks, list):
+            callbacks = [callbacks]
+
+        if not all(isinstance(callback, BaseCallback) for callback in callbacks):
+            raise TypeError(f"callbacks must be subclasses of BaseCallback.")
+
+        self._callbacks = callbacks
+
+    # XXX should be a method of MetaEstimatorMixin but this mixin can't handle all
+    # meta-estimators.
+    def _propagate_callbacks(self, sub_estimator, parent_node):
+        """Propagate the auto-propagated callbacks to a sub-estimator
+
+        Parameters
+        ----------
+        sub_estimator : estimator instance
+            The sub-estimator to propagate the callbacks to.
+
+        parent_node : ComputationNode instance
+            The computation node in this estimator to set as parent_node to the
+            computation tree of the sub-estimator. It must be the node where the fit
+            method of the sub-estimator is called.
+        """
+        if not hasattr(self, "_callbacks"):
+            return
+
+        if hasattr(sub_estimator, "_callbacks") and any(
+            isinstance(callback, AutoPropagatedMixin)
+            for callback in sub_estimator._callbacks
+        ):
+            bad_callbacks = [
+                callback.__class__.__name__
+                for callback in sub_estimator._callbacks
+                if isinstance(callback, AutoPropagatedMixin)
+            ]
+            raise TypeError(
+                f"The sub-estimators ({sub_estimator.__class__.__name__}) of a"
+                f" meta-estimator ({self.__class__.__name__}) can't have"
+                f" auto-propagated callbacks ({bad_callbacks})."
+                " Set them directly on the meta-estimator."
+            )
+
+        propagated_callbacks = [
+            callback
+            for callback in self._callbacks
+            if isinstance(callback, AutoPropagatedMixin)
+        ]
+
+        if not propagated_callbacks:
+            return
+
+        sub_estimator._parent_node = parent_node
+
+        if not hasattr(sub_estimator, "_callbacks"):
+            sub_estimator._callbacks = propagated_callbacks
+        else:
+            sub_estimator._callbacks.extend(propagated_callbacks)
+
+    def _eval_callbacks_on_fit_begin(self, *, levels, X=None, y=None):
```

Comment:
Would it be too magical to have only something call `_eval_callbacks_begin` and inspect internal the stack of calls to infer which method called this function.

Of course it would make sense only if the same methods are expected to be called for fit/predict/etc.

Reply:
> Of course it would make sense only if the same methods are expected to be called for fit/predict/etc.

Well that's not obvious at all and I've not really thought about that. This first iteration is all about fit. I think it will be easier to not try to be too magical for now